ci-operator/step-registry/baremetal/lab/post/wipe/baremetal-lab-post-wipe-commands.sh

```diff
@@ -64,24 +81,19 @@ function reset_host() {
   local pdu_uri="${7:-}"
   local host="${bmc_forwarded_port##1[0-9]}"
   host="${host##0}"
+  if ssh "${SSHOPTS[@]}" "root@${AUX_HOST}" ipmicmd "${host}" power status | grep -q "Error: Unable to establish IPMI v2" \
+  && [ -n "${pdu_uri}" ]; then
+    echo "Resetting ${host} pdu"
+    reset_pdu "${pdu_uri}"
+    sleep 30
+  fi
   timeout -s 9 10m ssh "${SSHOPTS[@]}" "root@${AUX_HOST}" prepare_host_for_boot "${host}" "pxe"
   if ! wait_for_power_down "$bmc_address" "$bmc_forwarded_port" "$bmc_user" "$bmc_pass" "$vendor" "$ipxe_via_vmedia"; then
```

Comment:
Your implementation is like an initial check and reboot if we fail to retrieve the status. I would be ok with it if:
1. we can use `timeout` before the ssh command
2. no sleep 30 -> some hardware may need more time, other less. Can we 'wait until the status is reported successfully (with timeout)'?

An alternative is to bind the timeout that fails in the if statement:

```suggestion
  if timeout -s 9 10m ssh "${SSHOPTS[@]}" "root@${AUX_HOST}" prepare_host_for_boot "${host}" "pxe" && wait_for_power_down "$bmc_address" "$bmc_forwarded_port" "$bmc_user" "$bmc_pass" "$vendor" "$ipxe_via_vmedia"; then
    echo "Cleaned up successfully"
  else 
```

Reply:
Regarding the suggestion: When the host is not reachable by IPMI, the issue is caused in `prepare_host_for_boot` on [line 38](https://github.com/openshift-qe/baremetal-qe-infra/blob/master/images/fcos-bastion-image/root/usr/bin/prepare_host_for_boot#L38) which will keep trying infinitely, not getting the expected string `Chassis Power is off`. In such case `timeout` is killed after 10m and when it is killed it even terminates the function so nothing else will execute in `reset_host()`. The `wait_for_power_down` function and everything else will not execute.

I have added timeout before the ssh and the `wait_for_power_down` for the check

Reply:
Oh, got it! We'd fix it too though. We should loop up to a maximum number of retries

Reply:
Also, wait_for_power_down may be not enough. If the server is set to restore the previous state when the power is back, it will not be off

Reply:
I think the until loop is fine as the host should not be unreachable. If it is, then it will be fixed with the pdu reboot in the previous step. We can remove the `wait_for_power_down` and add a sleep of 90-120, as the next step is for the until loop that will check for power off state for at least 10min. WDYT?

Reply:
I think it is prone to become an infinite loop... therefor having something like

For i in {1..X} max; do
 If I == max
   Break
 CheckPower
 Sleep Y
done

Reply:
Replaced `wait_for_power_down` with `sleep 90` and added PR to [Avoid infinite loop for node power off](https://github.com/openshift-qe/baremetal-qe-infra/pull/79)

Reply:
Can't the retry at line 49 suffice?

Reply:
Seems like line 49 would not do much. It does not have `$pdu_uri` passed to it. Besides this, there are other issues.
After `prepare_host_for_reboot` exits with failure with `reset_host()` execution first time, the `wait_for_power_down` would run for 21.30 minutes then will execute `reset_host()` second time without the pdu_uri (doesn't matter at this point). Timeout `prepare_host_for_reboot` fails again as pdu is not rebooted yet. `wait_for_power_down` runs for another 21.30 mins. After 43 mins the pdu is finally rebooted without the disk clean  up.

Reply:
Should we fix all of it then?

Reply:
The initial issue for which we tried twice wasn't related to the BMC unreachability, but rather to kernel panic events happening right while executing the cleanup. The only way for recovering was to reboot the BMC via PDU (because that BMC doesn't support any reboot endpoints via Redfish).

We need to maintain this behavior and add a resolution for the issue you're facing with the other vendor. 


Ideally, re-executing at most twice, anytime a 'fatal error' occurs (kernel panic, BMc fully unreachable, etc,...)

Reply:
I think this patch fixes the issues mentioned above. If a host is unreachable we would have to do a pdu reset for it to move forward. The node would be at least rebooted twice if there are any other issues.

Reply:
That's ok if we want to introduce this further reboot for the pdu, but while you're fixing this and given the analysis

> After prepare_host_for_reboot exits with failure with reset_host() execution first time, the wait_for_power_down would run for 21.30 minutes, then will execute reset_host() second time without the pdu_uri (doesn't matter at this point). Timeout prepare_host_for_reboot fails again as pdu is not rebooted yet. wait_for_power_down runs for another 21.30 mins. After 43 mins + 15 miins (of prepare_host_for_reboot running 2 times) the pdu is finally rebooted without the disk clean up.

Can we better set up the timeouts?